torchacc/dist/fsdp.py

```diff
@@ -202,3 +204,151 @@ def clip_grad_norm_(self, max_grad_norm):
 
     def forward(self, *args, **kwargs):
         return self.model(*args, **kwargs)
+
+    def optim_state_dict(self,
+                         optim: torch.optim.Optimizer,
+                         full_state_dict: str = 'FULL_STATE_DICT',
+                         rank0_only: bool = True,
+                         cpu_offload: bool = True) -> Dict[str, Any]:
+        # we only support FULL_STATE_DICT and flatten parameters now
+        assert full_state_dict, "FULL_STATE_DICT"
+        assert self.model.flatten_parameters, True
+        shard_meta_data = self.model.get_shard_metadata()
+
+        sharded_optim_state = optim.state_dict()['state']
+        optim_state_param_groups = optim.state_dict()['param_groups']
+        # the name of params to save is different
+        optim_state_param_groups[0]['params'].clear()
+        # unflattened and consolidated state_dict
+        consolidate_optim_state_dict: Dict[str, Any] = {
+            'state': {},
+            'param_groups': {}
+        }
+
+        if rank0_only and self.model.rank == 0:
+            consolidate_optim_state_dict[
+                'param_groups'] = optim_state_param_groups
+        if not rank0_only:
+            consolidate_optim_state_dict[
+                'param_groups'] = optim_state_param_groups
+
+        for layer_state, (layer_name, layer_params) in zip(
+                sharded_optim_state.values(),
+                shard_meta_data['flatten_info'].items()):
+            param_names, param_shapes, param_numels = layer_params
+            # get full_param_names of each layer in optim_state_dict
+            full_names = optim_utils._get_layer_full_names(
+                layer_name, param_names)
+            unflat_state_dict = {fn: {} for fn in full_names}
+
+            if rank0_only and self.model.rank == 0:
+                consolidate_optim_state_dict['param_groups'][0][
+                    'params'].append(full_names)
+            if not rank0_only:
+                consolidate_optim_state_dict[
+                    'param_groups'] = optim_state_param_groups
```

Comment:
why rank0 is different from others

Reply:
actually there's no difference, I made a mistake.